### PR TITLE
[CARBONDATA-3557] Write flink streaming data to partition table

### PR DESCRIPTION
**Pull request by niuge01**
Support write flink streaming data to a partition carbon table with stage file format.

 - [ ] Any interfaces changed?
Yes, add an property [COMMIT_THRESHOLD] of carbon writer.
 
 - [ ] Any backward compatibility impacted?
No
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1293/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1282/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1272/

File: core/src/main/java/org/apache/carbondata/core/statusmanager/StageInput.java [hunk 1]
```diff
@@ -39,6 +39,8 @@
    */
   private Map<String, Long> files;
 
+  private List<PartitionLocation> locations;
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/statusmanager/StageInput.java]: please add comment

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonS3Writer.java [hunk 2]
```diff
@@ -139,15 +152,16 @@ public void commit() throws IOException {
...
+      StageInput stageInput = this.uploadSegmentDataFiles(this.writePath, dataPath);
+      if (stageInput == null) {
+        return;
+      }
       try {
         String stageInputPath = CarbonTablePath.getStageDir(
             table.getAbsoluteTableIdentifier().getTablePath()) +
-            CarbonCommonConstants.FILE_SEPARATOR + this.writePartition;
-        StageManager.writeStageInput(stageInputPath, new StageInput(dataPath, fileList));
+            CarbonCommonConstants.FILE_SEPARATOR + UUID.randomUUID();// TODO UUID
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonS3Writer.java]: remove TODO

File: integration/flink/src/test/scala/org/apache/carbon/flink/TestSource.scala [hunk 3]
```diff
@@ -1,25 +1,27 @@
...
 
+import java.util.Random
+
 import org.apache.flink.api.common.state.{ListState, ListStateDescriptor}
 import org.apache.flink.runtime.state.{FunctionInitializationContext, FunctionSnapshotContext}
 import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction
 import org.apache.flink.streaming.api.functions.source.SourceFunction
 
-abstract class TestSource(val dataCount: Int) extends SourceFunction[String] with CheckpointedFunction {
+abstract class TestSource(val dataCount: Int) extends SourceFunction[Array[AnyRef]] with CheckpointedFunction {
```
Comment (jackylk) [inline integration/flink/src/test/scala/org/apache/carbon/flink/TestSource.scala]: Please add more test case to verify the write output is correct

File: core/src/main/java/org/apache/carbondata/core/statusmanager/StageInput.java (hunk 1 above)
Comment (niuge01) [reply core/src/main/java/org/apache/carbondata/core/statusmanager/StageInput.java]: done

File: integration/flink/src/test/scala/org/apache/carbon/flink/TestSource.scala (hunk 3 above)
Comment (niuge01) [reply integration/flink/src/test/scala/org/apache/carbon/flink/TestSource.scala]: done

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonS3Writer.java (hunk 2 above)
Comment (niuge01) [reply integration/flink/src/main/java/org/apache/carbon/flink/CarbonS3Writer.java]: done

**Comment by niuge01** [comment]
please test this

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1309/

**Comment by niuge01** [comment]
please test this

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1311/

**Comment by niuge01** [comment]
please test this

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1314/

**Comment by niuge01** [comment]
please test this

**Comment by niuge01** [comment]
please test this

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1337/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1327/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1320/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1343/

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonS3Writer.java [hunk 4]
```diff
@@ -184,29 +204,15 @@ public void close() {
...
-      try {
-        CarbonUtil.copyCarbonDataFileToCarbonStorePath(file.getAbsolutePath(), remotePath, 1024);
-      } catch (CarbonDataWriterException exception) {
-        LOGGER.error(exception.getMessage(), exception);
-        throw exception;
-      }
-      if (LOGGER.isDebugEnabled()) {
-        LOGGER.debug("Upload file[" + file.getAbsolutePath() + "] to [" + remotePath + "] end.");
-      }
+    final List<org.apache.carbondata.sdk.file.CarbonWriter> writers =
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonS3Writer.java]:
```suggestion
    final List<CarbonWriter> writers =
```

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java [hunk 5]
```diff
@@ -17,10 +17,229 @@
...
+
 /**
  * This class is a wrapper of CarbonWriter in SDK.
  * It not only write data to carbon with CarbonWriter in SDK, also generate segment file.
  */
-public abstract class CarbonWriter extends ProxyFileWriter<String> {
+public abstract class CarbonWriter extends ProxyFileWriter<Object[]> {
+
+  private static final Logger LOGGER =
+      LogServiceFactory.getLogService(CarbonS3Writer.class.getName());
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java]:
```suggestion
      LogServiceFactory.getLogService(CarbonWriter.class.getName());
```

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java [hunk 6]
```diff
@@ -17,10 +17,229 @@
...
+      uploadSegmentDataFiles(new File(localPath), remotePath, partitionLocationList, partitions);
+      if (partitionLocationList.isEmpty()) {
+        return null;
+      } else {
+        return new StageInput(remotePath, partitionLocationList);
+      }
+    }
+  }
+
+  private static void uploadSegmentDataFiles(
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java]: please add comment to describe this function

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java [hunk 7]
```diff
@@ -17,10 +17,229 @@
...
+    return this.identifier;
+  }
+
+  /**
+   * @return when there is no data file uploaded, then return <code>null</code>.
+   */
+  protected StageInput uploadSegmentDataFiles(final String localPath, final String remotePath) {
+    if (!this.table.isHivePartitionTable()) {
+      final File[] files = new File(localPath).listFiles();
+      if (files == null) {
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java]:
```suggestion
      if (files == null || files.length == 0) {
```

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java [hunk 8]
```diff
@@ -17,10 +17,229 @@
...
+      }
+      Map<String, Long> fileNameMapLength = new HashMap<>(files.length);
+      for (File file : files) {
+        fileNameMapLength.put(file.getName(), file.length());
+        if (LOGGER.isDebugEnabled()) {
+          LOGGER.debug(
+              "Upload file[" + file.getAbsolutePath() + "] to [" + remotePath + "] start.");
+        }
+        try {
+          CarbonUtil.copyCarbonDataFileToCarbonStorePath(file.getAbsolutePath(), remotePath, 1024);
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java]: create a constant for 1024

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java [hunk 9]
```diff
@@ -17,10 +17,229 @@
...
+        } catch (CarbonDataWriterException exception) {
+          LOGGER.error(exception.getMessage(), exception);
+          throw exception;
+        }
+        if (LOGGER.isDebugEnabled()) {
+          LOGGER.debug("Upload file[" + file.getAbsolutePath() + "] to [" + remotePath + "] end.");
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java]:
```suggestion
          LOGGER.debug("Upload file[" + file.getAbsolutePath() + "] to [" + remotePath + "] finished.");
```

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java [hunk 10]
```diff
@@ -17,10 +17,229 @@
...
+      if (partitionLocationList.isEmpty()) {
+        return null;
+      } else {
+        return new StageInput(remotePath, partitionLocationList);
+      }
+    }
+  }
+
+  private static void uploadSegmentDataFiles(
+      final File directory, final String remotePath,
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java]:
```suggestion
      final File localDirectory, final String remotePath,
```

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java [hunk 11]
```diff
@@ -17,10 +17,229 @@
...
+      final List<StageInput.PartitionLocation> partitionLocationList,
+      final List<String> partitions
+  ) {
+    final File[] files = directory.listFiles();
+    if (files == null) {
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java]:
```suggestion
    if (files == null || files.length == 0) {
```

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java [hunk 12]
```diff
@@ -17,10 +17,229 @@
...
+        LOGGER.debug("Upload file[" + file.getAbsolutePath() + "] to [" + remotePath + "] start.");
+      }
+      try {
+        CarbonUtil.copyCarbonDataFileToCarbonStorePath(file.getAbsolutePath(), remotePath, 1024);
+      } catch (CarbonDataWriterException exception) {
+        LOGGER.error(exception.getMessage(), exception);
+        throw exception;
+      }
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("Upload file[" + file.getAbsolutePath() + "] to [" + remotePath + "] end.");
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java]:
```suggestion
        LOGGER.debug("Upload file[" + file.getAbsolutePath() + "] to [" + remotePath + "] finished.");
```

File: integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java [hunk 13]
```diff
@@ -17,10 +17,229 @@
...
+
+    protected abstract org.apache.carbondata.sdk.file.CarbonWriter newWriter(final Object[] row);
+
+    public void reset() {
+      this.writers.clear();
+      this.root.children.clear();
+      this.root.writer = null;
+    }
+
+    private static final class Node {
```
Comment (jackylk) [inline integration/flink/src/main/java/org/apache/carbon/flink/CarbonWriter.java]: Please add comment, why is this required?

File: integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonPartitionWriter.scala [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+  val tableName = "test_flink_partition"
+
+  @Test
+  def testLocal(): Unit = {
+    sql(s"drop table if exists $tableName").collect()
+    sql(
+      s"""
+         | CREATE TABLE $tableName (stringField string, intField int, shortField short)
+         | STORED AS carbondata
+         | PARTITIONED BY (hour_ string, date_ string)
```
Comment (jackylk) [inline integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonPartitionWriter.scala]: can you add 2 more testcase: 
1. use Date type as partition column
2. use Int type as partition column
